module-system/sov-state/src/storage.rs

```diff
@@ -29,7 +30,21 @@ impl StorageKey {
         self.key.clone()
     }
 
-    pub fn as_cache_key(self) -> CacheKey {
+    pub fn as_cache_key(&self) -> &CacheKey {
+        debug_assert_eq!({
+            let StorageKey { key } = self;
+            key
+        }, {
+            let CacheKey { key } = unsafe { core::mem::transmute::<_, &CacheKey>(self) };
+            key
+        }, "the structure of the CacheKey must be exactly equal to the StorageKey so the unsafe transformation will work as expected");
+
+        // Safety: they are currently equivalent
+        // TODO https://github.com/Sovereign-Labs/sovereign-sdk/issues/643
+        unsafe { core::mem::transmute(self) }
+    }
+
+    pub fn to_cache_key(self) -> CacheKey {
         CacheKey { key: self.key }
     }
```

Comment:
This should be named `into_cache_key`: https://github.com/Swatinem/rust-cache/issues/162